File: MaximNigmatulin/Task4/src/main/kotlin/spbu/system/finegrained/FineGrainedSet.kt
package spbu.system.finegrained

import spbu.models.StudentRecord
import spbu.system.ExamSystemInterface

class FineGrainedSet : ExamSystemInterface {
    private val head = Node(
        next = null,
        key = -1,
        item = StudentRecord(-1, -1)
    )

    override fun add(record: StudentRecord): Boolean {
        val key = record.hashCode()
        head.lock()
        var pred = head
        try {
            if (pred.next == null) {
                pred.next = Node(null, key, record)
                return true
            }
            var curr = pred.next
            curr?.lock()
            try {
                while (curr != null && curr.key < key) {
                    pred.unlock()
                    pred = curr
                    curr = curr.next
                    curr?.lock()
                }
                if (curr != null && curr.key == key) {
                    return false
                }
                val insert = Node(curr, key, record)
                pred.next = insert
                return true
            } finally {
                curr?.unlock()
            }
        } finally {
            pred.unlock()
        }
    }

    override fun remove(record: StudentRecord): Boolean {
        val key = record.hashCode()
        head.lock()
        var pred = head
        try {
            if (pred.next == null) {
                pred.next = Node(null, key, record)
                return true
            }
            var curr = pred.next
            curr?.lock()
            try {
                while (curr != null && curr.key < key) {
                    pred.unlock()
                    pred = curr
                    curr = curr.next
                    curr?.lock()
                }
                if (curr != null && curr.key == key) {
                    pred.next = curr.next
                    return true
                }
                return false
            } finally {
                curr?.unlock()
            }
        } finally {
            pred.unlock()
        }
    }

    override fun contains(record: StudentRecord): Boolean {
        val key = record.hashCode()
        head.lock()
        var pred = head
        try {
            if (pred.next == null) {
                return false
            }
            var curr = pred.next
            curr?.lock()
            try {
                while (curr != null && curr.key < key) {
                    pred.unlock()
                    pred = curr
                    curr = curr.next
                    curr?.lock()
                }
                if (curr != null && curr.key == key) {
                    return true
                }
                return false
            } finally {
                curr?.unlock()
            }
        } finally {
            pred.unlock()
        }
    }

    override fun count(): Int {
        try {
            head.lock()
            var curr = head.next
            var count = 0
            while (curr != null) {
                curr.lock()
                curr = curr.next
                count += 1
            }
            return count
        } finally {
            var curr = head.next
            head.unlock()
            while (curr != null) {
                curr = curr.next
                curr?.unlock()
            }
        }
    }
}

File: KarimiHurmatullah/Task4/ExamSystem/ExamSystem/Collections/FineGrainedSet.cs
namespace ExamSystem.Collections
{
    public class FineGrainedSet<T>
    {
        private Node<T> head;
        private int count;

        public FineGrainedSet()
        {
            head = new Node<T>(int.MinValue);
            head.Next = new Node<T>(Int32.MaxValue);
        }

        public bool Add(T item)
        {
            int key = item.GetHashCode();
            head.Lock.Lock();
            Node<T> pred = head;
            try
            {
                Node<T> curr = pred.Next;
                curr.Lock.Lock();
                try
                {
                    while (curr.Key < key)
                    {
                        pred.Lock.Unlock();
                        pred = curr;
                        curr = curr.Next;
                        curr.Lock.Lock();
                    }
                    if (curr.Key == key)
                    {
                        return false;
                    }
                    Node<T> newNode = new Node<T>(item);
                    newNode.Next = curr;
                    pred.Next = newNode;
                    Interlocked.Increment(ref count);
                    return true;
                }
                finally
                {
                    curr.Lock.Unlock();
                }
            }
            finally
            {
                pred.Lock.Unlock();
            }
        }

        public bool Remove(T item)
        {
            Node<T> pred = null, curr = null;
            int key = item.GetHashCode();
            head.Lock.Lock();
            try
            {
                pred = head;
                curr = pred.Next;
                curr.Lock.Lock();
                try
                {
                    while (curr.Key < key)
                    {
                        pred.Lock.Unlock();
                        pred = curr;
                        curr = curr.Next;
                        curr.Lock.Lock();
                    }
                    if (curr.Key == key)
                    {
                        pred.Next = curr.Next;
                        Interlocked.Decrement(ref count);
                        return true;
                    }
                    return false;
                }
                finally
                {
                    curr.Lock.Unlock();
                }
            }
            finally
            {
                pred.Lock.Unlock();
            }
        }

        public bool Contains(T item)
        {
            Node<T> pred = null, curr = null;
            int key = item.GetHashCode();
            head.Lock.Lock();
            try
            {
                pred = head;
                curr = pred.Next;
                curr.Lock.Lock();
                try
                {
                    while (curr.Key < key)
                    {
                        pred.Lock.Unlock();
                        pred = curr;
                        curr = curr.Next;
                        curr.Lock.Lock();
                    }
                    return curr.Key == key;
                }
                finally
                {
                    curr.Lock.Unlock();
                }
            }
            finally
            {
                pred.Lock.Unlock();
            }
        }

        public int Count => count;
    }
}
